fix(collection): add options validation

kvwalker opened the pull request:
Fixes NODE-1671

This PR builds off of the recent options validation changes, which will be incorporated into `next` after the last PR (#1848) is merged.

mbroadst commented on lib/collection.js:
```diff
@@ -395,56 +402,37 @@ Collection.prototype.find = deprecateOptions(
-    // Translate to new command option noCursorTimeout
-    if (typeof newOptions.timeout === 'boolean') newOptions.noCursorTimeout = newOptions.timeout;
-
-    decorateCommand(findCommand, newOptions, ['session', 'collation']);
```
are we fully accounting for command decoration in the code? I see you have code over there to merge options from `finalOptions` into the command, but only if the key isn't `session`. Is that parity with this?

kvwalker replied:
Good catch

mbroadst commented on lib/collection.js:
```diff
@@ -340,9 +323,59 @@ Collection.prototype.find = deprecateOptions(
+        promiseLibrary: this.s.promiseLibrary,
+        ignoreUndefined: this.s.options.ignoreUndefined
+      },
+      { validationLevel: 'error' }
```
Let's default to `warn` for the first iteration here, as you said earlier people might accidentally be depending on broken options and we want to give them a chance to fix that rather than breaking their application.

Also, I would recommend setting this to a constant called like `LOG_DEFAULT` at the top of the file, so you can easily toggle between error modes during testing.

mbroadst commented on lib/collection.js:
```diff
@@ -522,6 +518,24 @@ Collection.prototype.insertMany = function(docs, options, callback) {
+    w: { type: ['number', 'string'] },
+    wtimeout: { type: 'number' },
+    j: { type: 'boolean' },
+    serializeFunctions: { type: 'boolean', default: this.s.serializeFunctions },
```
sometimes the default to `serializeFunctions` is `this.s.serializeFunctions`, and sometimes there is no default at all (L629). We should be consistent here I think

kvwalker replied:
I only set `this.s.serializeFunctions` as the default for `serializeFunctions` when it was [already being set](https://github.com/mongodb/node-mongodb-native/pull/1859/files/b77b26ac4403832c6e2aebc9d42319521bd6cdee#diff-a3644fc6e558da2bde71112d24ec75baL538). Should I be setting it as the default when it wasn't previously being assigned?

kvwalker replied:
Like we just talked about, I'll add this work to the follow-up epic and we can handle it then so no behavior is changed with this work.

mbroadst commented on lib/collection.js:
```diff
@@ -596,23 +607,40 @@ Collection.prototype.insertMany = function(docs, options, callback) {
+    serializeFunctions: { type: 'boolean' },
+    ordered: { type: 'boolean', default: true },
+    bypassDocumentValidation: { type: 'boolean' },
+    ignoreUndefined: { rejectUser: true },
```
Let's all bikeshed a better name for this, I think `rejectUser` isn't very clear. Maybe `overrideOnly`, `ignore<some other word>`

kvwalker commented on lib/collection.js:
```diff
@@ -353,38 +386,12 @@ Collection.prototype.find = deprecateOptions(
-    let newOptions = Object.assign({}, options);
-
-    // Make a shallow copy of the collection options
-    for (let key in this.s.options) {
```
I handled this by setting an override for `ignoreUndefined` (the only key in `mergeKeys`) [here](https://github.com/mongodb/node-mongodb-native/pull/1859/files/b77b26ac4403832c6e2aebc9d42319521bd6cdee#diff-a3644fc6e558da2bde71112d24ec75baR373).

kvwalker commented on lib/operations/collection_ops.js:
```diff
@@ -66,15 +66,9 @@ const groupFunction =
   // Create the bulk operation
   const bulk =
-    options.ordered === true || options.ordered == null
+    options.ordered === true
```
yep, that's right!

kvwalker commented on lib/options_validator.js:
```diff
@@ -132,7 +136,7 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
   });
 
-  return Object.freeze(verifiedOptions);
+  return verifiedOptions;
```
There are a couple things that don't currently work if we freeze the object so there's a ticket to do this in the follow-up epic: [NODE-1680](https://jira.mongodb.org/browse/NODE-1680)

kvwalker commented on test/unit/options_validator_tests.js:
```diff
@@ -292,8 +275,7 @@ describe('Options Validation', function() {
 
-    expect(validatedObject).to.deep.equal(testObject);
-    expect(validatedObject).to.be.frozen;
+    expect(validatedObject).to.deep.equal({ a: 'custom' });
```
Previously, I implemented override options to work more like defaults. I corrected that behavior so now an override will replace a provided option.

kvwalker commented on test/functional/find_tests.js:
```diff
@@ -2737,14 +2737,11 @@ describe('Find', function() {
-          client.close();
-          done();
-        });
+        expect(() => collection.find({}, options)).to.throw(
```
I agree that this is a superfluous test and we should delete it. @mbroadst, do you agree?

kvwalker commented on lib/options_validator.js:
```diff
@@ -78,11 +77,19 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
   Object.keys(optionsSchema).forEach(optionName => {
     const optionFromSchema = optionsSchema[optionName];
 
+    if (
```
I can definitely add documentation about how the schema compiler works. I can add it to the jsdocs for `validate`. Is there a better place to add it?

daprahamian replied:
I do not think there is a better place right now in repo. But as long as we have a place we can all see it, we can decide later on the best place to put it.

daprahamian commented on lib/options_validator.js:
```diff
@@ -108,10 +115,7 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
-            (optionType === 'object' && !isObject(optionValue)) ||
-            typeof optionValue !== optionType
-          ) {
+          if (typeof optionValue !== optionType) {
```
Would that not be covered by a special `array` type?

kvwalker commented on lib/options_validator.js:
```diff
@@ -71,18 +94,39 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
+      ) {
         console.warn(
           `A default value and override value were provided for option [${optionName}]. The override value will be used.`
         );
```
In adding the documentation for `validate()`, I realized that I am only [setting non-dynamic defaults](https://github.com/mongodb/node-mongodb-native/pull/1859/files#diff-e1fa59f7e4e4c9080c24be96453dfdb8R142) if `validationLevel !== VALIDATION_LEVEL_NONE`. It seems like I should be setting non-dynamic defaults regardless of the validation level. Also, deprecation warnings are only being emitted if the `validationLevel` is `warn` or `error`. I think that we should be emitting deprecation warnings regardless of validation level. It seems to me that the only thing the `validationLevel` should affect is type validation. What do you both think?

daprahamian replied:
I agree. Validation level should only affect type validation. Defaults should always be set, and deprecation warnings should always be emitted (and then handled by their respective node handlers)

mbroadst replied:
Also agree, can you please add some tests validating this behavior when you change it.

daprahamian commented on lib/collection.js:
```diff
@@ -280,7 +280,7 @@ const DEPRECATED_FIND_OPTIONS = ['maxScan', 'fields', 'snapshot'];
 const findSchema = {
   limit: { type: 'number', default: 0 },
-  sort: { type: ['array', 'object'] },
+  sort: { type: [Array, 'object'] },
```
Arrays are a special case in type comparisons. Instead of relying on `instanceof Array`, we should use `Array.isArray`. I think it would probably be a good idea to treat `'array'` as a special type at https://github.com/mongodb/node-mongodb-native/pull/1859/files#diff-e1fa59f7e4e4c9080c24be96453dfdb8R154 where we compare types. Something like:

```js
if (optionType === 'array') {
  if (!Array.isArray(optionValue)) {
    invalidate(validationLevel, invalidateMessageType, logger);
  }
}
```

mbroadst replied:
👍 with this idea

kvwalker commented:
With the last commit, I added tests for `validationLevel: 'none'` and changed the behavior of `validate` so it throws an error when a required option is missing, regardless of `validationLevel`.

kvwalker commented:
Closing this so we can look at a cleaner PR.
